### ISSUE TITLE
Help wanted: Query Expansion model finetuning — benchmarks, data balance, overfitting

### ISSUE THREAD

**Issue by tobi** [tobi/qmd#247]
## Context

QMD uses a small local model to expand raw user queries into our [Query Document syntax](https://github.com/tobi/qmd/blob/main/docs/SYNTAX.md). This is a critical part of the pipeline — it turns a bare query like `"UFC fight tonight"` into a structured multi-strategy search:

```
hyde: The Ultimate Fighting Championship held a Fight Night event featuring a main card bout with rankings implications and preliminary fights.
lex: UFC "Ultimate Fighting Championship" fight tonight
lex: MMA mixed martial arts event card
vec: Ultimate Fighting Championship fight event happening tonight
vec: UFC fight card and results for this evening
```

The expansion model generates three types of sub-queries:
- **`lex:`** — BM25 keyword variations (short, keyword-focused, preserves named entities)
- **`vec:`** — Semantic reformulations (natural language questions)  
- **`hyde:`** — Hypothetical document passage (~50-150 chars, what the answer might look like)

Full syntax spec: [docs/SYNTAX.md](https://github.com/tobi/qmd/blob/main/docs/SYNTAX.md)  
Scoring rubric: [finetune/SCORING.md](https://github.com/tobi/qmd/blob/main/finetune/SCORING.md)

---

## Architecture Overview

```
┌─────────────────────────────────────────────────────────────────┐
│                    QMD Inference Pipeline                         │
│                                                                   │
│  Raw Query ──→ [Expansion Model] ──→ Query Document               │
│                   (1.7B GGUF)         │                           │
│                                       ├─ lex: → BM25 Index ──┐   │
│                                       ├─ vec: → [Embedding] ─┤   │
│                                       └─ hyde:→ [Embedding] ─┤   │
│                                              (300M GGUF)      │   │
│                                                               ▼   │
│                                              RRF Fusion ──→ [Reranker] ──→ Results
│                                                              (0.6B GGUF)          │
└─────────────────────────────────────────────────────────────────┘
```

Three local GGUF models run the full pipeline — zero API calls:

| Component | Model | Size | Repo |
|-----------|-------|------|------|
| **Query Expansion** | Qwen3-1.7B (LoRA fine-tuned) | Q4_K_M | [`tobil/qmd-query-expansion-1.7B-gguf`](https://huggingface.co/tobil/qmd-query-expansion-1.7B-gguf) |
| **Embedding** | EmbeddingGemma-300M | Q8_0 | [`ggml-org/embeddinggemma-300M-GGUF`](https://huggingface.co/ggml-org/embeddinggemma-300M-GGUF) |
| **Reranker** | Qwen3-Reranker-0.6B | Q8_0 | [`ggml-org/Qwen3-Reranker-0.6B-Q8_0-GGUF`](https://huggingface.co/ggml-org/Qwen3-Reranker-0.6B-Q8_0-GGUF) |

**Future work (not yet fine-tuned):** The embedding and reranker models are off-the-shelf. Fine-tuning them on in-domain data would likely improve retrieval quality — contributors welcome.

---

## Data Creation Pipeline

Training data lives in [`finetune/data/*.jsonl`](https://github.com/tobi/qmd/tree/main/finetune/data). Currently ~13.5k examples from multiple sources:

### Data Sources

| Source | Script | Method | Quality | Count |
|--------|--------|--------|---------|-------|
| **Claude API** | [`dataset/generate_data.py`](https://github.com/tobi/qmd/blob/main/finetune/dataset/generate_data.py) | Claude generates expansions from diverse query templates | Highest | ~5k |
| **HF Datasets** | [`dataset/generate_data_offline.py`](https://github.com/tobi/qmd/blob/main/finetune/dataset/generate_data_offline.py) | Transforms [s-emanuilov/query-expansion](https://huggingface.co/datasets/s-emanuilov/query-expansion) dataset, adds synthetic hyde | Good | ~3k |
| **Ollama Local** | [`dataset/generate_ollama.py`](https://github.com/tobi/qmd/blob/main/finetune/dataset/generate_ollama.py) | Uses local GLM-4 via Ollama to generate expansions | Variable | ~2k |
| **GEPA Optimizer** | [`gepa/`](https://github.com/tobi/qmd/tree/main/finetune/gepa) | DSPy-based prompt optimization — uses teacher model to find optimal prompts, then generates at scale | High | ~1.5k |
| **Handcrafted** | `data/qmd_expansion_handcrafted.jsonl` | Manually written for edge cases, sports acronyms, domain-specific | Targeted | ~200 |
| **Sports Acronyms** | `data/qmd_expansion_sports.jsonl` | UFC, NFL, NBA, F1, IMSA, etc. — fixes acronym expansion failures | Targeted | 48 |

### Data Validation Pipeline

```
Raw JSONL files (data/*.jsonl)
  │
  ├─ validate_schema.py  — Ensures {"query": str, "output": [[type, text], ...]} format
  ├─ clean_data.py       — Fixes technical term issues, normalizes
  ├─ score_data.py       — Scores with reward.py rubric, flags low-quality
  │
  └─ prepare_data.py     — Converts to Qwen3 chat template, deduplicates, splits
     (or prepare_data_lfm2.py for LFM2 ChatML format)
       │
       └─ data/train/train.jsonl + val.jsonl (ready for training)
```

Each JSONL entry follows the [schema](https://github.com/tobi/qmd/blob/main/finetune/dataset/schema.py):
```json
{"query": "UFC fight tonight", "output": [["lex", "UFC Ultimate Fighting Championship fight tonight"], ["vec", "UFC fight event happening this evening"], ["hyde", "The Ultimate Fighting Championship..."]]}
```

---

## Training Pipeline

All training code: [`finetune/`](https://github.com/tobi/qmd/tree/main/finetune)  
Full docs: [`finetune/CLAUDE.md`](https://github.com/tobi/qmd/blob/main/finetune/CLAUDE.md)

### Stage 1: SFT (Supervised Fine-Tuning)

Standard LoRA fine-tuning on the prepared dataset.

| Setting | Value |
|---------|-------|
| Base model | Qwen3-1.7B or LFM2-1.2B |
| LoRA rank/alpha | 16 / 32 |
| Epochs | 5 |
| Batch size | 4 × 4 (gradient accumulation) |
| LR | 2e-4, cosine schedule |
| Config | [`configs/sft.yaml`](https://github.com/tobi/qmd/blob/main/finetune/configs/sft.yaml) (Qwen3), [`configs/sft_lfm2.yaml`](https://github.com/tobi/qmd/blob/main/finetune/configs/sft_lfm2.yaml) (LFM2) |

```bash
uv run dataset/prepare_data.py   # Prepare data first
uv run train.py sft --config configs/sft.yaml
```

### Stage 2: GRPO (Group Relative Policy Optimization)

RL fine-tuning on top of SFT using our [reward function](https://github.com/tobi/qmd/blob/main/finetune/reward.py) which scores:
- **Format compliance** (30 pts) — valid prefixes, no garbage
- **Diversity & coverage** (30 pts) — multiple types, no echo/duplication
- **Hyde quality** (20 pts) — concise, no repetition
- **Content quality** (20 pts) — relevance, lex=keywords, vec=natural language
- **Entity preservation** (20 pts, critical) — named entities must appear in lex lines

| Setting | Value |
|---------|-------|
| Base | SFT checkpoint |
| LoRA rank/alpha | 4 / 8 |
| Generations per prompt | 4 |
| KL beta | 0.04 |
| Max steps | 200 |
| Config | [`configs/grpo.yaml`](https://github.com/tobi/qmd/blob/main/finetune/configs/grpo.yaml) |

```bash
uv run train.py grpo --config configs/grpo.yaml
```

### Stage 3: GEPA (Generalized Efficient Prompt Adaptation)

DSPy-based prompt optimization for data generation (not model training). Uses a teacher model to find optimal few-shot prompts, then generates high-quality training data at scale.

```bash
# In finetune/gepa/
python optimizer.py    # Find optimal prompts
python generate.py     # Generate data using optimized prompts
```

### Deployment

```bash
uv run eval.py --model ./outputs/grpo -o eval_results.json   # Evaluate
uv run convert_gguf.py                                         # Merge + GGUF quantize
# Upload to HF only if eval improves
```

Can also train in the cloud via HuggingFace Jobs:
```bash
hf jobs uv run --flavor a10g-large --secrets HF_TOKEN --timeout 2h jobs/sft.py
```

### Supported Models

| Model | Status | Config |
|-------|--------|--------|
| **Qwen3-1.7B** | Primary, deployed | `sft.yaml` / `grpo.yaml` |
| **LFM2-1.2B** (Liquid) | Experimental | `sft_lfm2.yaml` |

---

## Current Problems

1. **Overfitting** — Models learn training distribution too well. Reproduce patterns rather than generalizing to novel queries. Especially visible with domain-specific queries (sports acronyms, niche topics).

2. **Unbalanced data** — The ~13.5k examples are heavily weighted toward tech/programming (generated via Claude). Underrepresented: sports, medical, legal, casual/conversational, non-English, ambiguous queries.

3. **No good benchmark** — This is the biggest gap. Standard RAG benchmarks (BEIR, MTEB) test retrieval quality but none map cleanly to our full pipeline (expand → lex+vec+hyde → BM25+vector fusion → rerank). We need a benchmark that tests whether the *expansion itself* improves retrieval.

4. **Embedding & reranker are off-the-shelf** — Only the expansion model is fine-tuned. Fine-tuning the embedding model (EmbeddingGemma-300M) and reranker (Qwen3-Reranker-0.6B) on in-domain data would complete the picture.

### Benchmark Design (highest priority)
- A benchmark measuring end-to-end retrieval quality: raw query → expansion → search → relevance
- Could build on existing datasets (NQ, HotpotQA, MS MARCO) with our pipeline as retrieval layer
- Should test: Does expanding with the model beat no expansion? Does model X beat model Y?
- Bonus: adversarial cases where expansion hurts (query drift)

### Data Quality & Balance
- More diverse training examples — especially non-tech domains
- Hard negatives — cases where the model should NOT over-expand  
- Short/ambiguous queries that need careful handling

### Training Improvements
- Better regularization for SFT → GRPO pipeline
- Curriculum learning ideas (easy → hard)
- Alternative reward signals beyond format compliance

### In-Domain Embedding & Reranker
- Fine-tuning EmbeddingGemma on QMD document collections
- Fine-tuning Qwen3-Reranker on QMD query/document pairs
- Would close the loop on fully in-domain retrieval

### General MLE Expertise
The current approach works but feels like we are plateauing. Fresh eyes from people who do this professionally would be incredibly valuable.

## Getting Started

```bash
git clone https://github.com/tobi/qmd.git
cd qmd/finetune

# Prepare data
uv run dataset/prepare_data.py

# Train SFT
uv run train.py sft --config configs/sft.yaml

# Evaluate
uv run eval.py --model ./outputs/sft

# Run reward function self-test
uv run reward.py
```

Happy to answer questions — drop a comment or open a PR. Even small improvements to training data or the scoring rubric help a lot.

**Comment by tobi** [comment]
A few more things worth calling out:

### Unused lex syntax features

The lex query syntax supports **quoted exact phrases** and **negation** but we have zero training data exercising them:

```
lex: "rate limiter" -redis        # exact phrase + negation
lex: "machine learning" -"deep learning"  # phrase exclusion
```

The expansion model never generates these because it has never seen them in training. This is a straightforward data gap — we need training examples that demonstrate when to use exact matching (proper nouns, multi-word terms) and when to exclude related-but-wrong terms.

### Entity stripping problem

The model has a bad habit of dropping named entities during expansion. Example:

**Query:** `meeting with Bob about C++`
**Bad expansion (current behavior):**
```
lex: c++ meetings
vec: programming meeting notes
```

Bob is completely gone. This should score very low — and our [entity preservation scoring](https://github.com/tobi/qmd/blob/main/finetune/SCORING.md#5-named-entity-preservation-0-20-points-critical) already penalizes this (-30 heavy penalty when no lex lines contain entities), but the model still does it. Likely because the training data is dominated by impersonal technical queries where there are no personal names to preserve.

**Good expansion should be:**
```
lex: Bob "C++" meeting
vec: meeting notes with Bob about C++ programming
hyde: Bob and I discussed C++ implementation details during our meeting, covering template metaprogramming and memory management.
```

### Summary of low-hanging fruit for contributors

1. **Add quoted phrase examples** to training data — when the query contains multi-word proper nouns, the lex output should quote them
2. **Add negation examples** — queries where excluding related terms improves precision
3. **Add personal/named entity queries** — meetings, emails, conversations with people, project names
4. **Improve reward function** — boost entity preservation penalty weight, or add specific detection for name-dropping
5. **In-domain embedding & reranker fine-tuning** — currently off-the-shelf models, would complete the fully in-domain pipeline

Lots of fun stuff to dig into. Each of these is a self-contained contribution.

**Comment by qdrddr** [comment]
It makes sense to allow all the models via API: Reranker, Embedding, and Query Expansion.
This may be especially useful for the GEPA prompt optimizer to optimize the prompt with a strong SotA model available via API and then transfer the optimized prompt for Query Expansion.

https://github.com/tobi/qmd/issues/114

**Comment by tonze** [comment]
Tried qmd for the first time today on my own hybrid search setup. Noticed something that might be useful: Not all query expansions help equally in RRF:

• Lexical expansions add the most value closing the vocabulary gap
• Vector expanions are okay, adds more signal to good matches
• HyDE is hit or miss. Brilliant for topical queries ("agentic coding" → fake summary lands right in the cluster of real talks). But harmful for entity queries ("Werner Vogels" → generates generic "tech CTO talks about cloud" that matches everything and dilutes the actual name signal)

The tradeoff: every expansion adds noise. If HyDE buries a good result below the reranker cutoff, it's gone.

Just an observation from testing. Might be worth weighing expansion types differently.

**Comment by homanp** [comment]
Some thoughts on the datasets:

**Domain breakdown (3,579 in-repo examples):**
- 74% "other/general" (mostly homogeneous "how to..." / "what is..." queries)
- 17% tech/programming
- 2.5% sports, 1.8% health, 1.5% finance, 1.5% cooking, 1.2% travel, 0.4% legal

**Structural gaps:**
- Only 15 examples with negation/exclusion (0.4%) — model can't learn `-term` in lex
- Only 15% have named entities, despite entity preservation being scored at -65pt penalty
- Question-form queries are over-represented (34%) vs bare keywords and conversational queries

The "reproduces patterns rather than generalizing" symptom is textbook for large-but-homogeneous data.

**What I'd prioritize:**

1. **Domain balance**: equal representation across sports, medical, legal, finance, shopping, local/geographic, casual. Currently these are all < 3% each.

2. **Query structure mix**: real searches are mostly bare keywords and keyword phrases, not well-formed questions. Target: 20% bare keywords, 25% keyword phrases, 20% questions, 10% conversational/vague, 10% entity-heavy, 5% negation, 5% time-sensitive, 5% `/only:` modifiers.

3. **Expansion difficulty scaling**: this is the biggest gap. Need explicit hard cases where expansion should be *conservative*: already-precise queries (`"CVE-2024-31497"`), navigational queries (`"github login"`), polysemous queries (`"rust"`, `"ACL"`), and adversarial query-drift bait (`"apple stock price"` must not expand to fruit).

4. **Seed from real query distributions**: (MS MARCO, Natural Questions, Google Trends) instead of LLM templates.

5. **Hard negative mining**: run current model on diverse queries, find failures, use corrected outputs as training data.

Happy to help build out data for any of these slices.

**Comment by zacharypodbela** [comment]
I'd be interested in helping to design a benchmark of end-to-end retrieval quality. Here's how I would think about approaching this - open to feedback and input before executing on this.

### Part 1: Generating a varied collection of personal documents to serve as our test corpus

**Option 1:** Outfit an Existing Dataset

Unsurprisingly, people don't publish their personal notes so there aren't a ton of great data sets to mimic the kind of collections people might actually use with QMD. 

I did find:
- This dataset of [500k real emails from Enron employees](https://www.kaggle.com/datasets/wcukierski/enron-email-dataset) which were released by federal regulators. We could have Claude rewrite these emails as markdown meeting notes, summaries, project docs, etc. (The emails are actually organized into folders by mailbox owner, allowing Claude to "live" the life of a specific person day-by-day.)
- This dataset of [100 hours of recorded meetings with transcripts](https://huggingface.co/datasets/knkarthick/AMI). These seem especially valuable given the rise of meeting note takers as a source of documentation (and therefore as a use case for QMD).

**Option 2:** Generate Completely Synthetic Data

We could give Claude personas and contexts (Ex: "You're a senior engineer at a mid-size startup. Generate 30 meeting notes from the past 6 months with realistic names, projects, decisions, and action items.") and ask it to generate documents. We'd need to ensure multiple docs mention the same people, projects, and topics, so we're testing cases where multiple documents are relevant to a search.

**Tradeoffs**

I think we should prioritize Option 1 as it seems more important to benchmark against real-world examples of how humans are likely to communicate (real names, real projects, real ambiguity). That being said, the advantage of prioritizing Option 2 is that once we have the foundational prompts/scripts for synthetic data generation, if we find gaps in our dataset, extending the library is as easy as asking Claude to role-play as a new character.

### Part 2: Creating a set of queries + relevance judgments

We create a script that chooses passages at random, and asks Claude to create a query from that passage. We could create an array of "styles" of queries we want to make sure to cover and have Claude generate X examples of each:

| Category | Example | Tests |
|----------|---------|-------|
| Bare keywords | `oauth migration` | Basic lex retrieval |
| Natural language | `"what was the decision on auth"` | vec/hyde value |
| Entity-heavy | `"meeting with Bob and Sarah"` | Entity preservation |
| Ambiguous/short | `"deploy"` | Expansion quality |
| Already precise | `"CVE-2024-31497"` | Conservative expansion |

**Additional Complexity / Notes**

- We'd expect some of these backward generated queries will match multiple documents, and we'll need to figure out how to find all the relevant documents (not just the one used to generate the query). We can solve this with pooled relevance judgments: run multiple retrieval methods against our document set, take the union of their top-K results, then review that pool to label which documents are actually relevant.
- Queries generated by Claude from passages may end up being easier than real queries because Claude has the answer text right in front of it and will naturally use similar vocabulary. I think this is find for v1, but in the future we might consider making sure there is coverage for things like mis-spellings or close queries (simulating mis-remembering a few of the details on a real document).
- For simplicity, in v1 I think we just label documents as just "relevant / not relevant" (binary). We'll be able to compute MRR and Recall@K but not nDCG (which needs graded judgments like "highly relevant / somewhat relevant / not relevant").

**Comment by rkbadhan** [comment]
Hi @tobi 
I have pushed a small change to tighten the reward design specifically for handling the Compound entity chaining, Filler word penalty in lex lines,  Stricter diversity threshold, Broaden quoted-phrase bonus. Please review the PR - https://github.com/tobi/qmd/pull/256

**Comment by tobi** [comment]
Closing as part of the post-v2.5.1 backlog cleanup: this item is over two months old and has either been superseded by newer QMD releases or is too stale to keep actionable. If it still reproduces on v2.5.1 or newer, please open a fresh focused issue with current repro steps.
